lib/webmock/util/hash_counter.rb

```diff
@@ -5,22 +5,36 @@
 module WebMock
   module Util
     class HashCounter
-      attr_accessor :hash
+      attr_accessor :hash, :array
 
       def initialize
         self.hash = Hash.new(0)
         @order = {}
         @max = 0
         @lock = ::Mutex.new
+        self.array = []
+        @request_object_ids = {}
       end
 
       def put(key, num=1)
         @lock.synchronize do
+          store_to_array(key:, num:)
           hash[key] += num
           @order[key] = @max += 1
         end
       end
 
+      def store_to_array(key:, num:)
+        request_object_id = @request_object_ids[key]
+        request_object_id = key.object_id if request_object_id.nil?
+        num.times do
+          array << ObjectSpace._id2ref(request_object_id)
+        rescue RangeError
+          # Points to an invalid or recycled object so ignore
+        end
+        @request_object_ids[key] = key.object_id
+      end
```

Comment:
@bblimke this is missing any comments to explain what's going on - maybe worth me adding some?

For others - doing this so that we are just referencing existing requests in the array, rather than creating new objects.

This is so that if we have massive numbers of requests that have the same signature, we're not bloating memory.

Reply:
@johngallagher Thanks for your work on this. I've been looking at the `store_to_array` method.

I think it might be helpful to add a comment explaining the intention behind using `object_id` and `_id2ref`. This would really help to understand the reasoning behind this approach.

That said, I believe we might be able to simplify this code. I'm not sure whether we really need `object_id` and `_id2ref` at all. From my understanding, `array` is just storing references to the original objects, not clones.

I think the following achieves the same result:

```ruby
def store_to_array(key, num)
  @request_objects ||= {}
  stored_object = @request_objects[key] ||= key
  num.times do
    array << stored_object
  end
end
```

This version uses a hash (`@request_objects`) to store the original request signature objects with same `#hash` result. It maintains the original object without the need for `object_id`.

It should be more efficient as it avoids the overhead of `ObjectSpace._id2ref` and potential `RangeError` rescues,
though I don't know how much faster that is.

Can you think of any scenarios where the original implementation behaves differently from this proposed one?

What are your thoughts on this?

I wonder if this can me optimised even further.

Reply:
This is a **much** nicer solution than mine! I can't think of any scenarios just now but I'll think on it over the next few days...